File: MSHFView.m
#import "public/MSHFView.h"

@implementation MSHFView

BOOL boost;

- (instancetype)initWithFrame:(CGRect)frame {
  self = [self initWithFrame:frame
                 audioSource:[[MSHFAudioSourceASS alloc] init]];
  return self;
}

- (instancetype)initWithFrame:(CGRect)frame
                  audioSource:(MSHFAudioSource *)audioSource {
  self = [super initWithFrame:frame];

  if (self) {
    self.numberOfPoints = 8;
    self.waveOffset = 0;
    self.gain = 0;
    self.limiter = 0;
    self.sensitivity = 1;
    self.disableBatterySaver = false;
    self.autoHide = true;
    MSHFHidden = self.autoHide;

    if (self.autoHide) {
      [self setAlpha:0.0f];
    } else {
      [self setAlpha:1.0f];
    }

    self.audioSource = audioSource;
    self.audioSource.delegate = self;

    self.audioProcessing = [[MSHFAudioProcessing alloc] initWithBufferSize:1024];
    self.audioProcessing.delegate = self;
    self.audioProcessing.fft = true;

    [self initializeWaveLayers];

    cachedLength = self.numberOfPoints;
    self.points = (CGPoint *)malloc(sizeof(CGPoint) * self.numberOfPoints);

    NSDictionary *prefs = [[NSDictionary alloc] initWithContentsOfFile:MSHFPrefsFile];
    boost = ([prefs objectForKey:@"MSHFAirpodsSensBoost"] ? [[prefs objectForKey:@"MSHFAirpodsSensBoost"] boolValue] : NO);
  }

  return self;
}

-(void)dealloc {
    [_displayLink invalidate];
    free(self.points);
}

- (void)setAutoHide:(BOOL)value {
  if (value &&
      (silentSince < ((long long)[[NSDate date] timeIntervalSince1970] - 1))) {
    MSHFHidden = true;
    [self setAlpha:0.0f];
  } else {
    MSHFHidden = false;
    [self setAlpha:1.0f];
  }

  _autoHide = value;
}

- (void)stop {
    if (self.audioSource.isRunning && !self.disableBatterySaver) {
        [self.audioSource stop];
        [self.displayLink setPaused:true];
        silentSince = -2;
        [self redraw];
    }
}

- (void)start {
  NSString *identifier = [[NSProcessInfo processInfo] processName];
  if ([identifier isEqualToString:@"Music"] || [identifier isEqualToString:@"Spotify"] || [[NSClassFromString(@"SBMediaController") sharedInstance] isPlaying]) {
	    [self.audioSource start];
      [self.displayLink setPaused:false];
  }
}

- (void)initializeWaveLayers {
}

- (void)resetWaveLayers {
}

- (void)configureDisplayLink {
    
  self.displayLink = [CADisplayLink displayLinkWithTarget:self
                                                 selector:@selector(redraw)];

  [self.displayLink addToRunLoop:[NSRunLoop currentRunLoop]
                         forMode:NSDefaultRunLoopMode];
  [self.displayLink setPaused:true];

  self.displayLink.preferredFramesPerSecond = 60;
}

- (void)updateWaveColor:(UIColor *)waveColor
           subwaveColor:(UIColor *)subwaveColor {
}

- (void)updateWaveColor:(UIColor *)waveColor
           subwaveColor:(UIColor *)subwaveColor
        subSubwaveColor:(UIColor *)subSubwaveColor {
}

- (void)redraw {
  if (self.autoHide) {
    if (silentSince < ((long long)[[NSDate date] timeIntervalSinceReferenceDate] - 1)) {
      if (!MSHFHidden) {
        MSHFHidden = true;
        [UIView animateWithDuration:0.5
                         animations:^{
                           [self setAlpha:0.0f];
                         }];
      }
    } else if (MSHFHidden) {
      MSHFHidden = false;
      [UIView animateWithDuration:0.5
                       animations:^{
                         [self setAlpha:1.0f];
                       }];
    }
  }
}

- (void)updateBuffer:(float *)bufferData withLength:(int)length {
  if (self.autoHide) {
    for (int i = 0; i < length / 4; i++) {
      if (bufferData[i] > 0.000005) {
        silentSince = (long long)[[NSDate date] timeIntervalSinceReferenceDate];
        break;
      }
    }
  }

  [self.audioProcessing process:bufferData withLength:length];
}

- (void)setSampleData:(float *)data length:(int)length {
  NSUInteger const compressionRate = length / self.numberOfPoints;

  float const pixelFixer = self.bounds.size.width / self.numberOfPoints;

  if (cachedLength != self.numberOfPoints) {
    free(self.points);
    self.points = (CGPoint *)malloc(sizeof(CGPoint) * self.numberOfPoints);
    cachedLength = self.numberOfPoints;
  }


  if ([[[NSClassFromString(@"BluetoothManager") sharedInstance] connectedDevices] count] && boost) {

    for (int i = 0; i < self.numberOfPoints; i++) {
        self.points[i].x = i * pixelFixer;
        double pureValue = data[i * compressionRate] * self.gain;

        if (!pureValue) {
          self.points[i].y = self.waveOffset;
          continue;
        }

        if (self.limiter != 0) {
          pureValue = (fabs(pureValue) < self.limiter
                          ? pureValue
                          : (pureValue < 0 ? -1 * self.limiter : self.limiter));
        }

        self.points[i].y = (pureValue * self.sensitivity);

        while (fabs(self.points[i].y) < 1.5) {
            self.points[i].y *= 25;
        }
        self.points[i].y += self.waveOffset;
      }
  } else {

    for (int i = 0; i < self.numberOfPoints; i++) {
      self.points[i].x = i * pixelFixer;
      double pureValue = data[i * compressionRate] * self.gain;

      if (self.limiter != 0) {
        pureValue = (fabs(pureValue) < self.limiter
                        ? pureValue
                        : (pureValue < 0 ? -1 * self.limiter : self.limiter));
      }

      self.points[i].y = (pureValue * self.sensitivity) + self.waveOffset;

      if (isnan(self.points[i].y))
        self.points[i].y = self.waveOffset;
    }
  }
}

@end

File: MSHFView.swift
import UIKit

private var boost:Bool = false

@objc (MSHFView) public class MSHFView: UIView, MSHFAudioDelegate, MSHFAudioProcessingDelegate {
    private var cachedLength = 0
    private var silentSince: TimeInterval = 0
    private var MSHFHidden = false

    @objc private var shouldUpdate = false
    @objc internal var disableBatterySaver = false
    private var _autoHide = true
    @objc internal var autoHide: Bool {
        get {
            _autoHide
        }
        set(value) {
            if value && (silentSince < (Date().timeIntervalSince1970 - 1)) {
                MSHFHidden = true
                alpha = 0.0
            } else {
                MSHFHidden = false
                alpha = 1.0
            }
        }
    }
    
    @objc internal var numberOfPoints = 0
    @objc internal var gain: CGFloat = 0.0
    @objc internal var limiter: CGFloat = 0.0
    @objc internal var waveOffset: CGFloat = 0.0
    @objc internal var sensitivity: CGFloat = 0.0
    @objc internal var displayLink: CADisplayLink?
    internal var points: UnsafeMutablePointer<CGPoint>!
    internal var siriEnabled = false
    internal var waveColor: UIColor?
    internal var subwaveColor: UIColor?
    internal var subSubwaveColor: UIColor?
    private var audioSource: MSHFAudioSource?
    @objc internal var audioProcessing: MSHFAudioProcessing?

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }

    override convenience init(frame: CGRect) {
      self.init(frame: frame, audioSource: MSHFAudioSourceASS())
    }

    public init(frame: CGRect, audioSource: MSHFAudioSource?) {

      super.init(frame: frame)

      numberOfPoints = 8
      waveOffset = 0
      gain = 0
      limiter = 0
      sensitivity = 1
      disableBatterySaver = false
      autoHide = false
      MSHFHidden = autoHide

      if autoHide {
          self.alpha = 0.0
      } else {
          self.alpha = 1.0
      }

      if let audioSource = audioSource {
          self.audioSource = audioSource
      }
      self.audioSource?.delegate = self

      audioProcessing = MSHFAudioProcessing(bufferSize: 1024)
      audioProcessing!.delegate = self
      audioProcessing!.fft = true

      initializeWaveLayers()

      shouldUpdate = true

      cachedLength = numberOfPoints
      points = unsafeBitCast(malloc(MemoryLayout<CGPoint>.size * numberOfPoints), to: UnsafeMutablePointer<CGPoint>.self)

      let MSHFPrefsFile = "/var/mobile/Library/Preferences/com.ryannair05.mitsuhaforever.plist"
      if let prefs = NSDictionary(contentsOfFile:MSHFPrefsFile) {
          boost = prefs["MSHFAirpodsSensBoost"] as? Bool ?? false
      }
    }

  @objc public func stop() {
    if audioSource?.isRunning ?? true && !disableBatterySaver {
        audioSource?.stop()
    }
  }

  @objc public func start() {
    let identifier = ProcessInfo.processInfo.processName
    if (identifier == "Music") || (identifier == "Spotify") || ((NSClassFromString("SBMediaController")?.sharedInstance().isPlaying()) == true) || FileManager.default.fileExists(atPath: "/Library/MobileSubstrate/DynamicLibraries/RoadRunner.dylib") {
        audioSource?.start()
     }
  }

  internal func initializeWaveLayers() {

  }

  internal func resetWaveLayers() {

  }

  internal func configureDisplayLink() {
    displayLink = CADisplayLink(target: self, selector: #selector(redraw))

    displayLink!.add(to: RunLoop.current,forMode: .default)
    displayLink!.isPaused = false

    displayLink!.preferredFramesPerSecond = 60
  }

  @objc public func updateWaveColor(_ waveColor: UIColor,subwaveColor: UIColor) {

  }

  @objc public func updateWaveColor(_ waveColor: UIColor,subwaveColor: UIColor,subSubwaveColor: UIColor) {

  }

  @objc internal func redraw() {
       if autoHide {
          if silentSince < (Date().timeIntervalSince1970 - 1) {
              if MSHFHidden {
                  MSHFHidden = false
                  UIView.animate(
                      withDuration: 0.5,
                      animations: {
                          self.alpha = 0.0
                      })
              }
          } else if !MSHFHidden {
              MSHFHidden = true
              UIView.animate(
                  withDuration: 0.5,
                  animations: {
                      self.alpha = 1.0
                })
          }
      } 
    }

   public func updateBuffer(_ bufferData: UnsafeMutablePointer<Float>, withLength length: Int) {
      if autoHide {
          for i in 0..<(length / 4) {
              if bufferData[i] > 0.000005 || bufferData[i] < -0.000005 {
                  silentSince = Date().timeIntervalSince1970
                  break
              }
          }
      }

      audioProcessing?.process(bufferData, withLength: length)
  }

  internal func setSampleData(_ data: UnsafeMutablePointer<Float>?, length: Int) {
      let compressionRate = length / numberOfPoints
      let pixelFixer: CGFloat = bounds.size.width / CGFloat(numberOfPoints)

      if cachedLength != numberOfPoints {
          free(points)
          points =  unsafeBitCast(malloc(MemoryLayout<CGPoint>.size * numberOfPoints), to: UnsafeMutablePointer<CGPoint>.self)
          cachedLength = numberOfPoints
      }

      if boost {

          for i in 0..<numberOfPoints {
              points[i].x = CGFloat(i) * pixelFixer
              var pureValue: CGFloat = CGFloat(data![i * compressionRate]) * gain

              if pureValue == 0.0 {
                  points[i].y = waveOffset
                  continue
              }

              if limiter != 0 {
                  pureValue = abs(pureValue) < limiter
                      ? pureValue
                      : (pureValue < 0 ? -1 * limiter : limiter)
              }

              points[i].y = (pureValue * sensitivity)

              while abs(points[i].y) < 1.5 {
                  points[i].y *= 25
              }
              points[i].y += waveOffset
          }
      }
      else {
        for i in 0..<numberOfPoints {
          points[i].x = CGFloat(i) * pixelFixer
          var pureValue: CGFloat = CGFloat(data![i * compressionRate]) * gain

          if limiter != 0 {
              pureValue = abs(pureValue) < limiter
                  ? pureValue
                  : (pureValue < 0 ? -1 * limiter : limiter)
          }

          points[i].y = (pureValue * sensitivity) + waveOffset

          if points[i].y.isNaN {
              points[i].y = waveOffset
          }
        }
      }
    }
}